README
libcmdline: a command line parsing library
------------------------------------------

The libcmdline is a command line parsing library written in c++.
It supports single character short options (e.g. -v, -n, ...), long gnu-style
options (e.g. --verbose, ...) and positional parameters (e.g. list of 
filenames given to a command). 
libcmdline automatically generates a help message, that can be displayed
via the "--help" option.
It is also possible to display a "HTMLified" help message with "--usageHTML"
and an "XML" help message (used for frontend programs) with "--usageXML".

How to use it?
--------------

1. Create some command line arguments (CmdArgs).

CmdArgs represent all the options that can be given to a program on the 
command line. There are several types of command line options:

CmdArgSwitch: an option without a parameter
CmdArgType<TYPE>: an option with a parameter of type TYPE
CmdArgVector<TYPE>: an option that takes a list of parameters of type TYPE
CmdArgThrow<E>: an option that throws an exception E when specified on the
	command line (used for --help).


Examples:

<code>
CmdArgType<int> count('c', "count", "HOW_MANY_FOOS", "select number of foos.");
// <int> -> parameter type given as a c++ template parameter
// 'c' -> short option name
// "count" -> long option name
// "HOW_MANY_FOOS" -> parameter name
// "select ..." -> option description
count.setDefaultValue(42);
</code>

This creates an optional command line option with a required integer 
parameter (named HOW_MANY_FOOS). On the command line it can be specified in 
the standard fashions:
"-c17", "-c 17", "--count 17" or "--count=17".

If the integer parameter is not given on the command line an error message is
displayed.

If the count option is not specified on the command line, its value is set to
the supplied default value.

The parsed value can be retrieved by "count.value()". 


<code>
CmdArgSwitch verbose('v', "verbose", "display verbose messages");
</code>

This creates an option without a parameter.

If it is specified on the command line by "-v" or "--verbose", its value
("verbose.value()") is true, otherwise it is false.



A positional parameter can be created by omitting the short and long option
names:

<code>
CmdArgType<std::string> outfile("FILENAME", "specify the output file");
</code>



By default options are optional. You can change this by setting the 
"CmdArg::isREQ" flag of an option as the last parameter of its constructor:

<code>
CmdArgType<std::string> infile("FILENAME", "specify the input file", CmdArg::isREQ);
</code>

If a required option is not given on the command line, and error message is 
displayed.

2. Create an argument iterator:
An argument iterator is used to retrieve the strings given to the parser.
If you want to parse strings given on the command line, you should use an 
"ArgvIter" and give it the c-style array containing the command line 
parameters:

<code>
int main(int argc, char** argv)
{
  ...
  ArgvIter iter(--argc, ++argv);
  ...
}
</code>


It is also possible to get the strings from a file or (more generaly) from an
input stream. You can do this via an "IStreamIter".


3. Create a command line object.

The CmdLine class is responsible for all the parsing work.

<code>
//          /program name\  /  short description  \
CmdLine cmd("testprogram", "this is a test program");
// add long description
cmd.description("This is the powerful test program. You can use it to ...");
</code>



4. Add the command line option objects the the CmdLine object.

Options are added to the CmdLine object by its "append" method. Calls to
CmdLine::append should be placed in a try/catch block since append throws 
CmdLineErrors if you append an option that uses already used short/long option
names.

<code>
try
{
  cmd.append(&count);
  cmd.append(&verbose);
  cmd.append(&infile);
  cmd.append(&outfile);
}
catch (const CmdLineError& e)
{
  std::cerr << "Error: " << e.str() << std::endl;
  exit(1);
}
</code>



5. Start the parsing.

<code>
try
{
  cmd.parse(iter);
}
catch (const CmdLineUsageError& e) // --help specified
{
  // display usage message
  cmd.usage()
  exit(0);
}
catch (const CmdLineSyntaxError& e)  // invalid command line
{
  cmd.error() << "invalid command line arguments:" << std::endl
              << e.str() << std::endl;
  exit(-1);
}
catch (const CmdLineUsageHTMLError& e) // --usageHTML specified
{
  cmd.usageHTML(std::cerr);
  exit(0);
}
catch (const CmdLineUsageXMLError& e) // --usageXML specified
{
  cmd.usageXML(std::cerr);
  exit(0);
}
</code>


6. Process the parsed values.

The parsed value of an option can be retried via the "value()" method.
You can check if an optional option was given with the "CmdArg::given()" 
method.

<code>
std::cout << "count=" << count.value << std::endl;
if (verbose.value()) std::cerr << "verbose mode" << std::endl;
std::cerr << "infile=" << infile.value() << std::endl;
if (outfile.given()) std::cerr << "outfile=" << outfile.value() << std::endl;
</code>



The whole example code:

<code>
#include <iostream>

#include "ArgvIter.hh"
#include "CmdArgs.hh"
#include "CmdLine.hh"
#include "Compiler.hh"
#include "Errors.hh"

int main(int argc, char** argv)
{
  // options
  CmdArgType<int> count('c', "count", "HOW_MANY_FOOS", "number of foos.");
  count.setDefaultValue(42);	
	
  CmdArgSwitch verbose('v', "verbose", "display verbose messages");

  CmdArgType<std::string> infile("FILENAME", "specify the input file", 
                                 CmdArg::isREQ);

  CmdArgType<std::string> outfile("FILENAME", "specify the output file");

  // argument iterator
  ArgvIter iter(--argc, ++argv);

  // CmdLine object
  CmdLine cmd("testprogram", "this is a test program");
  cmd.description("This is the powerful test program. You can use it to ...");

  // add options
  try
  {	
    cmd.append(&count);
    cmd.append(&verbose);
    cmd.append(&infile);
    cmd.append(&outfile);
  }
  catch (const CmdLineError& e)
  {
    std::cerr << "Error: " << e.str() << std::endl;
    exit(1);
  }	

  // parse and handle exceptions
  try
  {
    cmd.parse(iter);
  }
  catch (const CmdLineUsageError& e) // --help specified
  {
    // display usage message
    cmd.usage()
    exit(0);
  }
  catch (const CmdLineSyntaxError& e)  // invalid command line
  {
    cmd.error() << "invalid command line arguments:" << std::endl
                << e.str() << std::endl;
    exit(-1);
  }
  catch (const CmdLineUsageHTMLError& e) // --usageHTML specified
  {
    cmd.usageHTML(std::cerr);
    exit(0);
  }
  catch (const CmdLineUsageXMLError& e) // --usageXML specified
  {
    cmd.usageXML(std::cerr);
    exit(0);
  }

  // process parsed values
  std::cout << "count=" << count.value << std::endl;
  if (verbose.value()) std::cerr << "verbose mode" << std::endl;
  std::cerr << "infile=" << infile.value() << std::endl;
  if (outfile.given()) std::cerr << "outfile=" << outfile.value() << std::endl;

  return 0;
}

debian/changelog
libcmdline (0.1.0-1) unstable; urgency=low

  * Initial release

 -- Joerg Mechnich <mechnich@informatik.uni-freiburg.de>  Fri, 19 May 2006 22:36:10 +0200



src/test/testCmdLine.cc
/**************************************************************************
**       Title: test program for libCmdLine that uses LMBUNIT
**    $RCSfile: testCmdLine.cc,v $
**   $Revision: 1.2 $$Name:  $
**       $Date: 2002/04/11 08:45:14 $
**   Copyright: GPL $Author: ronneber $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: testCmdLine.cc,v $
**  Revision 1.2  2002/04/11 08:45:14  ronneber
**  - adapted to new position of lmbunit.h
**  - added check for new Compiler<float>
**
**  Revision 1.1  2002/04/04 12:32:45  pigorsch
**  - initial revision
**
**
**
**************************************************************************/

#include "lmbunit.hh"

#include <ArgvIter.hh>
#include <CmdLine.hh>
#include <CmdArgs.hh>
#include <Compiler.hh>
#include <Errors.hh>

static void testConstructorDefaultHelp()
{
  CmdLine cmd("test", "test program");
  char* argv="--help";
  ArgvIter ai(1, &argv);
  try
  {
    cmd.parse(ai);
    LMBUNIT_WRITE_FAILURE("expected excetion 'CmdLineUsageError'");
  }
  catch (CmdLineUsageError e)
  {
    return;
  }
}

static void testConstructorDefaultUsageHTML()
{
  CmdLine cmd("test", "test program");
  char* argv="--usagehtml";
  ArgvIter ai(1, &argv);
  try
  {
    cmd.parse(ai);
    LMBUNIT_WRITE_FAILURE("expected excetion 'CmdLineUsageHTMLError'");
  }
  catch (CmdLineUsageHTMLError e)
  {
    return;
  }
}

static void testConstructorDirectAppending()
{
  CmdArgType<int> arg1('a', "arg1", "value", "this is arg1");
  CmdArgType<int> arg2('b', "arg2", "value", "this is arg2");
  CmdLine cmd("test", "test program", true, &arg1, &arg2, 0);
    
  const char* a1="--arg1";
  const char* a2="42";
  const char* a3="-b";
  const char* a4="17";
  const char* const argv[]={a1, a2, a3, a4};

  ArgvIter ai(4, argv);
  cmd.parse(ai);
  return;
}

static void testDuplicateArgs()
{
  CmdArgType<int> arg1('a', "arg1", "value", "this is arg1");
  CmdArgType<int> arg2('b', "Arg1", "value", "this is arg2");
  try
  {
    CmdLine cmd("test", "test program", true, &arg1, &arg2, 0);
    LMBUNIT_WRITE_FAILURE("expected excetion 'CmdLineDublicateArgError'");
  }
  catch (CmdLineDublicateArgError e)
  {
    return;
  }
}

static void testFloatArg()
{
  CmdArgType<float> miau('p', "pi", "value", 
                         "give your own definition of pi");
  
  CmdLine cmd("test", "test program", true, &miau, NULL);
  

  const char* argv[] = {"-p", "3.1415"};
  ArgvIter ai( 2, argv);
  cmd.parse( ai);
  
  LMBUNIT_ASSERT_EQUAL_DELTA( miau.value(), 3.1415, 0.0000001);
}



int main()
{
  LMBUNIT_WRITE_HEADER();
  
  LMBUNIT_RUN_TEST(testConstructorDefaultHelp());
  LMBUNIT_RUN_TEST(testConstructorDefaultUsageHTML());
  LMBUNIT_RUN_TEST(testConstructorDirectAppending());
  LMBUNIT_RUN_TEST(testDuplicateArgs());
  LMBUNIT_RUN_TEST(testFloatArg());

  LMBUNIT_WRITE_STATISTICS();
  
  return _nFails;
}



src/test/test_CmdLine.cc
/**************************************************************************
**       Title: test program for libcmdline
**    $RCSfile: test_CmdLine.cc,v $
**   $Revision: 1.6 $$Name:  $
**       $Date: 2002/04/22 15:58:36 $
**   Copyright: GPL $Author: pigorsch $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: test_CmdLine.cc,v $
**  Revision 1.6  2002/04/22 15:58:36  pigorsch
**  - uses "deafult" functions
**
**  Revision 1.5  2002/04/11 08:45:26  ronneber
**  - adapted to new position of lmbunit.h
**
**  Revision 1.4  2002/04/10 11:08:09  pigorsch
**  - testing IStreamIter :-)
**
**  Revision 1.3  2002/04/04 12:36:41  pigorsch
**  - constructor changed again
**
**  Revision 1.2  2002/03/27 14:36:07  ronneber
**  - adapted to new CmdLine constructor
**
**  Revision 1.1  2002/03/26 07:45:29  ronneber
**  restructuring for autoconf
**
**  Revision 1.1  2002/03/26 07:36:28  ronneber
**  restructuring for autoconf
**
**  Revision 1.1.1.1  2002/03/22 13:45:07  pigorsch
**  moved from polsoft repository
**
**
**
**************************************************************************/

#include <iostream>

#include "IStreamIter.hh"
#include "ArgvIter.hh"
#include "CmdArgs.hh"
#include "CmdLine.hh"
#include "Compiler.hh"
#include "Errors.hh"

#include <cstdlib>

int main(
    int argc,
    char** argv)
{
  CmdArgSwitch toggle1('a', "toggle1", "This toggles feature a");
  
  CmdArgSwitch toggle2('b', "toggle2", "This activates feature b");
  
  CmdArgType<int> count(
      'c', "count", "number", "Number of hurzes.");
  count.setDefaultValue(0);
  
  CmdArgTypeVector<std::string> files(
      'f', "files", "filenames", "Files to delete.");
  
  CmdArgTypeFixedVector<double> position(
      'p', "position", "coordinates", "The initial 3dim position", 3);
  std::vector<double> def;
  def.push_back(1);
  def.push_back(2);
  def.push_back(3);
  position.setDefaultValue(def);
  
  CmdArgType<std::string> image(
      'i', "image", "filename", 
      "Image to show. The specified image file must exist and "
      "must be accessible by the current user. This sentence is "
      "so long because we want to known if line breaking works.");
  image.setDefaultValue("hurz.png");
  
  CmdArgType<int> answer(
      0, "answer", "life_universe_everything", 
      "Should be 42.\nWow - Look at that indentation...");
  answer.setDefaultValue(42);
  
  CmdArgTypeVector<std::string> hurzes(
      "hurzes", "All the hurzes we need. This is used for "
      "checking if vectors work...",
      CmdArg::Syntax(CmdArg::isOPT|CmdArg::isVALOPT));
  //CmdArg::isREQ);
  
  
  CmdLine cmd(argv[0], "just a test program", true);

  cmd.description(
      "Once you realize that documentation should be laughed at, peed upon, "
      "put on fire, and just ridiculed in general, THEN, and only then, have "
      "you reached the level where you can safely read it and try to use it "
      "to actually implement a driver.\n\n"
      "	                                 -- Linus Torvalds");
  
  ArgvIter iter(--argc, ++argv);
  //IStreamIter iter(std::cin);

  try
  { 
    cmd.append(&toggle1);
    cmd.append(&toggle2);
    cmd.append(&count);
    cmd.append(&files);
    cmd.append(&position);
    cmd.append(&image);
    cmd.append(&answer);
    cmd.append(&hurzes);
    
    cmd.parse(iter);
  }
  catch (CmdLineUsageError e)
  {
    cmd.usage();
    std::exit(0);
  }
  catch (CmdLineUsageHTMLError e)
  {
    cmd.usageHTML(std::cerr);
    std::exit(0);
  }
  catch (CmdLineSyntaxError e)
  {
    cmd.error() << "Oops: you seem to have given invalid command line arguments...\n"
                << e.str() << std::endl;
    cmd.usage();
    std::exit(-1);
  }
  catch (CmdLineError e)
  {
    cmd.error() << "Oops: an CmdLine error occured. That's bad...\n"
                << e.str() << std::endl;
    std::exit(-2);
  }
  catch (...)
  {
    cmd.error() << "Oops: an unknown error occured. That's really bad...\n";
    std::exit(-3);
  }

  std::cerr << "count   =" << count.value() << std::endl
       << "toggle1 =" << (toggle1.value() ? "true" : "false") << std::endl
       << "toggle2 =" << (toggle2.value() ? "true" : "false") << std::endl
       << "image   =" << image.value() << std::endl
       << "answer  =" << answer.value() << std::endl
       << "position=[";

  for (CmdArgTypeVector<double>::const_iterator p=position.begin();
       p!=position.end();
       ++p)
  {
    std::cerr << *p << " ";
  }
  std::cerr << "]" << std::endl;

  std::cerr << "hurzes  =[";
  
  for (CmdArgTypeVector<std::string>::const_iterator h=hurzes.begin();
       h!=hurzes.end();
       ++h)
  {
    std::cerr << *h << " ";
  }
  std::cerr << "]" << std::endl;  
}


src/libcmdline/Compiler.cc
/**************************************************************************
**       Title: compilers for various datatypes
**    $RCSfile: Compiler.cc,v $
**   $Revision: 1.1 $$Name:  $
**       $Date: 2002/04/11 08:42:01 $
**   Copyright: GPL $Author: ronneber $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: Compiler.cc,v $
**  Revision 1.1  2002/04/11 08:42:01  ronneber
**  - moved Compilers.hh to Compiler.cc (template specializations must be in a .cc file!)
**
**  Revision 1.1  2002/03/26 07:36:28  ronneber
**  restructuring for autoconf
**
**  Revision 1.1.1.1  2002/03/22 13:45:07  pigorsch
**  moved from polsoft repository
**
**
**
**************************************************************************/

#ifndef COMPILERS_HH
#define COMPILERS_HH

#include <cctype>
#include <cstdlib>
#include <iostream>

#include "Compiler.hh"

template<> 
bool 
Compiler<int>::compile(
    const char* arg,
    const char*& endptr,
    int& value)
{
  if (arg==0) // null string - nothing to do
  {
    return true; 
  } 
  else if (*arg==0) // empty string
  {
    setErrorMsg("empty integer value specified");
    return false;
  }
  
  // compile the string into an integer
  char* ptr;
  int result=strtol(arg, &ptr, 0);  
  
  if (ptr==arg) // first char is invalid
  {
    setErrorMsg("invalid integer value \""+std::string(arg)+"\"");
    return false;
  }
  
  endptr=ptr;
  value=(int)result;
  return true;
};

template<> 
bool 
Compiler<float>::compile(
    const char* arg,
    const char*& endptr,
    float& value)
{
  if (arg==0) // null string - nothing to do
  {
    return true; 
  } 
  else if (*arg==0) // empty string
  {
    setErrorMsg("empty integer value specified");
    return false;
  }
  
  // compile the string into an integer
  char* ptr;
  float result=strtod(arg, &ptr);  
  
  if (ptr==arg) // first char is invalid
  {
    setErrorMsg("invalid float value \""+std::string(arg)+"\"");
    return false;
  }
  
  endptr=ptr;
  value=result;
  return true;
};


template<> 
bool 
Compiler<double>::compile(
    const char* arg,
    const char*& endptr,
    double& value)
{
  if (arg==0) // null string - nothing to do
  {
    return true; 
  } 
  else if (*arg==0) // empty string
  {
    setErrorMsg("empty integer value specified");
    return false;
  }
  
  // compile the string into an integer
  char* ptr;
  double result=strtod(arg, &ptr);  
  
  if (ptr==arg) // first char is invalid
  {
    setErrorMsg("invalid double value \""+std::string(arg)+"\"");
    return false;
  }
  
  endptr=ptr;
  value=result;
  return true;
};

template<> 
bool 
Compiler<std::string>::compile(
    const char* arg,
    const char*& endptr,
    std::string& value)
{
  if (arg!=0)
  {
    endptr=0;
    value=arg;
  } 

  return true;
}; 

template<> 
bool 
Compiler<bool>::compile(
    const char* arg,
    const char*& endptr,
    bool& value)
{
  if (arg==0) // no argument -> true
  {
    value=true;
    return true; 
  } 
  else if (*arg==0) // empty string -> true
  {
    value=true;
    return true;
  }

  std::string s(arg); 
  for (std::string::iterator p=s.begin();
       p!=s.end();
       ++p)
  {
    *p=tolower(*p);
  }
  
  if (s=="1" || s=="on" || s=="true")
  {
    endptr=0;
    value=true;
    return true;
  }
  else if (s=="0" || s=="off" || s=="false")
  {
    endptr=0;
    value=false;
    return true;
  }
  
  return false;
};

#endif


src/libcmdline/CmdArg.cc
/**************************************************************************
**       Title: basic command line argument
**    $RCSfile: CmdArg.cc,v $
**   $Revision: 1.7 $$Name:  $
**       $Date: 2005/06/09 07:57:12 $
**   Copyright: GPL $Author: tschmidt $
** Description:
**
**    
**
**-------------------------------------------------------------------------
**
**  $Log: CmdArg.cc,v $
**  Revision 1.7  2005/06/09 07:57:12  tschmidt
**  - Changed enum names from bla to CmdLinebla
**
**  Revision 1.6  2002/06/03 15:23:52  pigorsch
**  - optional parameters don't have []'s in descriptions anymore
**
**  Revision 1.5  2002/05/27 15:36:41  pigorsch
**  - removed parameter "detailed" in function synopsis
**
**  Revision 1.4  2002/04/22 15:56:57  pigorsch
**  - removed some comments
**
**  Revision 1.3  2002/04/18 13:55:03  pigorsch
**  - removed parameter fixedSize from constructors
**  - changed way of clearing flags
**
**  Revision 1.2  2002/04/10 09:54:20  pigorsch
**  - constructors set pModified=false
**  - adjustSyntax now throws syntax error exception on error
**
**  Revision 1.1  2002/03/26 07:36:28  ronneber
**  restructuring for autoconf
**
**  Revision 1.1.1.1  2002/03/22 13:45:07  pigorsch
**  moved from polsoft repository
**
**
**
**************************************************************************/

#include "CmdArg.hh"
#include "CmdLine.hh"

CmdArg::
CmdArg(
    char shortName,
    const std::string longName,
    const std::string valueName,
    const std::string description,
    CmdArg::Syntax syntax)
        :pFlags((CmdArg::CmdLineFlags)0),
         pSyntax(syntax),
         pSequence(0),
         pShortName(shortName),
         pLongName(longName),
         pValueName(valueName),
         pDescription(description),
         pModified(false)
{
  adjustSyntax();
}

CmdArg::
CmdArg(
    char shortName,
    const std::string longName,
    const std::string description,
    CmdArg::Syntax syntax)
        :pFlags((CmdArg::CmdLineFlags)0),
         pSyntax(syntax),
         pSequence(0),
         pShortName(shortName),
         pLongName(longName),
         pValueName(),
         pDescription(description),
         pModified(false)       
{
  adjustSyntax();
}

CmdArg::
CmdArg(
    const std::string valueName,
    const std::string description,
    CmdArg::Syntax syntax)
        :pFlags((CmdArg::CmdLineFlags)0),
         pSyntax(syntax),
         pSequence(0),
         pShortName(0),
         pLongName(),
         pValueName(valueName),
         pDescription(description),
         pModified(false)
{

  adjustSyntax();
}

CmdArg::
~CmdArg()
{}

void
CmdArg::
adjustSyntax()
{
  // If the value is specified as both OPTIONAL and REQUIRED
  // then assume it is required.
  //
  if (hasSyntax(isVALREQ) && hasSyntax(isVALOPT)) 
  {
    pSyntax=CmdArg::Syntax(syntax() & (~isVALOPT));
  }

  // If they said the argument was both STICKY and SEPARATE then
  // I dont know what to think just just ignore both of them.
  //
  if (hasSyntax(isVALSTICKY) && hasSyntax(isVALSEP)) 
  {
    pSyntax=CmdArg::Syntax(syntax() & (~(isVALSTICKY | isVALSEP)));
  }

  // If a non-NULL, non-empty value-name was given but we werent
  // told that the argument takes a value, then assume that it
  // does take a value and that the value is required.
  //
  if (!pValueName.empty() && (!hasSyntax(isVALTAKEN))) 
  {
    pSyntax=CmdArg::Syntax(syntax() | isVALREQ);
  }

  // If a value is taken and the argument is positional, then
  // we need to make isREQ and isOPT consistent with isVALREQ
  // and isVALOPT
  //
  if (hasSyntax(isVALTAKEN) && hasSyntax(isPOS)) 
  {
    pSyntax=CmdArg::Syntax(syntax() & (~(isREQ | isOPT)));
    if (hasSyntax(isVALREQ)) pSyntax=CmdArg::Syntax(syntax() | isREQ);
    if (hasSyntax(isVALOPT)) pSyntax=CmdArg::Syntax(syntax() | isOPT);
  }
  
  // If a value is taken but no value name was given,
  // then default the value name.
  //
  if (hasSyntax(isVALTAKEN) && (pValueName.empty())) 
  {
    pValueName="value";
  }

  // If argument has fixed size>0 then it must me a fixed list.
  if (fixedSize()>0)
  {
    pSyntax=CmdArg::Syntax(syntax() | isFIXEDLIST);
    
    // if an argument is a fixed list, it cant be a normal list.
    pSyntax=CmdArg::Syntax(syntax() & (~isLIST));
  }
  else
  {
    pSyntax=CmdArg::Syntax(syntax() & (~isFIXEDLIST));
  }
  
  // If no keyword name or character name was given, then the
  // argument had better take a value and it must be positional
  //
  if ((!pShortName) && (pLongName.empty()) 
      && (!hasSyntax(isPOS))) 
  {
    if (hasSyntax(isVALTAKEN))
    {
      pSyntax=CmdArg::Syntax(syntax() | isPOS);
    } 
    else 
    { 
      CmdLineSyntaxError se;
      se << "syntax definition error: non-positional CmdArg has no short or long name";
      throw(se);
    }
  }
}

const std::string
CmdArg::
synopsis(
    bool displayShort,
    bool displayLong) const
{
  char sn=shortName();
  std::string ln=longName();
  std::string vn=valueName();
          
  std::ostringstream ost;

  if (!hasSyntax(CmdArg::isPOS)) 
  {
    if (ln.empty()) 
    {
      displayLong=false;
    }
    if (sn==0) 
    {
      displayShort=false;
    }
            
    if (!displayShort && !displayLong)
    {
      if (sn!=0)
      {
        displayShort=true;
      }
      else if (!ln.empty())
      {
        displayLong=true;
      }
    }
  }
          
  //if (!hasSyntax(CmdArg::isREQ)) ost << '[';
          
  if (displayShort && displayLong
      && !hasSyntax(CmdArg::isPOS) 
      && hasSyntax(CmdArg::isVALTAKEN) 
      && hasSyntax(CmdArg::isVALSTICKY))
  {
    if (hasSyntax(CmdArg::isVALOPT))
    {
      ost << '-' << char(sn) << '[' << vn << "]|--"
          << ln << "[=" << vn << ']';
    } 
    else 
    {
      ost << '-' << sn << vn << "|--" << ln << '=' << vn;
    }
            
    if (hasSyntax(CmdArg::isFIXEDLIST))
    {
      ost << '(' << fixedSize() << ')';
    }            

    if (hasSyntax(CmdArg::isLIST)) 
    {
      ost << " ...";
    }

    //if (!hasSyntax(CmdArg::isREQ))
    //{
    //  ost << ']';
    //}

    return ost.str();
  }
          
  if (!hasSyntax(CmdArg::isPOS)) 
  {
    if (displayShort)
    {
      ost << '-' << char(sn);
      if (displayLong)
      {
        ost << '|';
      }
    }
    if (displayLong)
    {
      ost << "--" << ln;
    }
                    
    if (hasSyntax(CmdArg::isVALTAKEN))
    {
      if (!hasSyntax(CmdArg::isVALSTICKY)) 
      {
        ost << ' ';
      }
    }
  }
          
  // If the argument takes a value then print the value
  if (hasSyntax(CmdArg::isVALTAKEN))
  {
    if (!hasSyntax(CmdArg::isPOS) &&
        hasSyntax(CmdArg::isVALOPT))
    {
      ost << '[';
    }
    if (hasSyntax(CmdArg::isVALSTICKY)) 
    {
      if (!displayShort && displayLong) ost << '=';
    }
    ost << vn;

    if (hasSyntax(CmdArg::isFIXEDLIST))
    {
      ost << '(' << fixedSize() << ')';
    }
    
    if (hasSyntax(CmdArg::isLIST)) 
    {
      ost << " ...";
    }
    if (!hasSyntax(CmdArg::isPOS) &&
        hasSyntax(CmdArg::isVALOPT))
    {
      ost << ']';
    }
  }
          
  //if (!hasSyntax(CmdArg::isREQ)) 
  //{
  //  ost << ']';
  //}

  return ost.str();
}
